### DISCUSSION TITLE
Change body shape and(or) it's properties

### DISCUSSION THREAD

**Discussion by wrenge** [bepu/bepuphysics2#324] [Q&A] [answered]
Hello! Couldn't find anywhere on the Internet how to change properties of existing shape. I think I'm figured it out, but I'm not sure  is it the right approach.
```cs
TypedIndex originalShape = Sim.Bodies.GetDescription(colRef.BodyHandle).Collidable.Shape;
var shape = Sim.Shapes.GetShape<Box>(originalShape.Index);
shape.Width = opts.Size.X;
shape.Height = opts.Size.Y;
shape.Length = opts.Size.Z;
inertia = shape.ComputeInertia(opts.Mass);

var desc = Sim.Bodies.GetDescription(colRef.BodyHandle);
desc.LocalInertia = inertia;
```
However I don't know how change shape type (for example box to capsule) without completely recreating body. Is it even intended?

**Comment by RossNordby** [comment, marked as answer]
Almost! `GetShape` returns a reference to the shape. Since the shape itself is a value type, assigning to a non-ref variable just copies it, so it's no longer a reference to the original memory.

If you changed it to:
```
ref var shape = ref Sim.Shapes.GetShape<Box>(originalShape.Index);
```
then it would be a memory reference.

`GetDescription` works differently; it builds a `BodyDescription` for you, but modifying it will not modify the original because it's just a copy of the data again. Because there is no place in the library where a `BodyDescription` actually sits around, it can't return a `ref` to it. You can use `ApplyDescription` to set all of a body's properties to match the description.

You can also grab a `BodyReference` by doing `Simulation.Bodies[bodyHandle]`. `BodyReference` has a bunch of handy stuff on it, including a `LocalInertia` property which *does* grab a direct reference to the local inertia data for the body. There are some subtleties to using direct memory references, though; if the body is asleep, changing the memory directly doesn't wake it up (you'd have to use `BodyReference.Awake = true`). Also, directly changing the inertia tensor memory won't handle transitions between dynamic and kinematic by itself. 

`BodyReference.SetLocalInertia` lets you specify a new local inertia and it'll handle the transitions for you.

If you want to change the shape referenced by a body, you can use `BodyReference.SetShape`. Like `SetLocalInertia`, it'll handle transitions and awakening. Note that you don't have to use that if you've just modified the shape memory directly; the body is still pointing at the same memory either way. You just might have to call `Awake= true` if you want it to wake up to respond to new collisions or something.

**Reply by wrenge** [reply]
Thanks for the reply!
Do I need to destroy old shape when I'm creating new for existing object?
